src/storage.rs

```diff
@@ -126,136 +154,173 @@ impl MemStorageCore {
         cs: ConfState,
         pending_membership_change: Option<(ConfState, u64)>,
     ) {
-        self.snapshot.mut_metadata().set_conf_state(cs);
+        self.raft_state.conf_state = cs;
         if let Some((cs, idx)) = pending_membership_change {
-            self.snapshot
-                .mut_metadata()
-                .set_pending_membership_change(cs);
-            self.snapshot
-                .mut_metadata()
-                .set_pending_membership_change_index(idx);
+            self.raft_state.pending_conf_state = Some(cs);
+            self.raft_state.pending_conf_state_start_index = Some(idx);
+        }
+    }
+
+    /// Save raft logs and apply configuration changes in them.
+    pub fn append(&mut self, ents: &[Entry]) -> Result<()> {
+        if ents.is_empty() {
+            return Ok(());
+        }
+
+        // Gaps are not allowed in raft logs.
+        if self.entries.last().unwrap().get_index() + 1 < ents[0].get_index() {
+            return Err(Error::Store(StorageError::LogGap));
+        }
+
+        // Remove all entries overwritten by `ents`, and truncate `ents` if need.
+        let next_idx = self.entries.last().unwrap().get_index() + 1;
+        let first_idx = cmp::max(ents[0].get_index(), self.entries[0].get_index());
+        if next_idx > first_idx {
+            let diff = next_idx - first_idx;
+            // Uncommitted raft logs can't be compacted.
+            assert!(self.entries.len() >= diff as usize);
+            let s = self.entries.len() - diff as usize;
+            while self.entries.len() > s {
+                self.entries.pop();
+            }
+        }
+
+        let i = ents
+            .iter()
+            .position(|e| e.get_index() == first_idx)
+            .unwrap();
+        ents[i..].iter().for_each(|e| self.entries.push(e.clone()));
+        Ok(())
+    }
+
+    fn inner_first_index(&self) -> u64 {
+        let first_index = self.entries[0].get_index();
+        if self.snapshot_metadata.get_index() == first_index {
+            // The first entry's index equals to the snapshot index means the entry is "dummy"
```

Comment:
How about removing the dummy entry completely?

Reply:
I have tried. It proves that we need to set `snapshot_metadata` to a correct value in some cases. To avoid change those test cases, I prefer to keep it.